src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -50,18 +51,174 @@
+
+    (void)GRPC_ERROR_REF(error);
+
+    ac->mu_.Lock();
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+      gpr_log(GPR_INFO, "CLIENT_CONNECT: %s: on_writable: error=%s",
+              ac->addr_str_.c_str(), grpc_error_std_string(error).c_str());
+    }
+    GPR_ASSERT(ac->fd_);
+    fd = ac->fd_;
+    ac->fd_ = nullptr;
+    ac->mu_.Unlock();
+
+    grpc_timer_cancel(&ac->alarm_);
+
+    ac->mu_.Lock();
+    if (error != GRPC_ERROR_NONE) {
+      error = grpc_error_set_str(error, GRPC_ERROR_STR_OS_ERROR,
+                                 "Timeout occurred");
+      goto finish;
+    }
+
+    do {
+      so_error_size = sizeof(so_error);
+      err = getsockopt(grpc_fd_wrapped_fd(fd), SOL_SOCKET, SO_ERROR, &so_error,
+                       &so_error_size);
+    } while (err < 0 && errno == EINTR);
+    if (err < 0) {
+      error = GRPC_OS_ERROR(errno, "getsockopt");
+      goto finish;
+    }
+
+    switch (so_error) {
+      case 0:
+        grpc_pollset_set_del_fd(ac->interested_parties_, fd);
+        *ep = grpc_tcp_client_create_from_fd(fd, ac->channel_args_,
+                                             ac->addr_str_);
+        fd = nullptr;
+        break;
+      case ENOBUFS:
+        /* We will get one of these errors if we have run out of
+           memory in the kernel for the data structures allocated
+           when you connect a socket.  If this happens it is very
+           likely that if we wait a little bit then try again the
+           connection will work (since other programs or this
+           program will close their network connections and free up
+           memory).  This does _not_ indicate that there is anything
+           wrong with the server we are connecting to, this is a
+           local problem.
+
+           If you are looking at this code, then chances are that
+           your program or another program on the same computer
+           opened too many network connections.  The "easy" fix:
+           don't do that! */
+        gpr_log(GPR_ERROR, "kernel out of buffers");
+        ac->mu_.Unlock();
+        grpc_fd_notify_on_write(fd, &ac->write_closure_);
+        return;
+      case ECONNREFUSED:
+        /* This error shouldn't happen for anything other than connect(). */
+        error = GRPC_OS_ERROR(so_error, "connect");
+        break;
+      default:
+        /* We don't really know which syscall triggered the problem here,
+           so punt by reporting getsockopt(). */
+        error = GRPC_OS_ERROR(so_error, "getsockopt(SO_ERROR)");
+        break;
+    }
+
+  finish:
+    if (fd != nullptr) {
+      grpc_pollset_set_del_fd(ac->interested_parties_, fd);
+      grpc_fd_orphan(fd, nullptr, nullptr, "tcp_client_orphan");
+      fd = nullptr;
+    }
+    done = (--ac->refs_ == 0);
+    ac->mu_.Unlock();
+    if (error != GRPC_ERROR_NONE) {
+      std::string str;
+      bool ret = grpc_error_get_str(error, GRPC_ERROR_STR_DESCRIPTION, &str);
+      GPR_ASSERT(ret);
```

Comment:
I think this can just be `GPR_ASSERT(grpc_error_get_str(error, GRPC_ERROR_STR_DESCRIPTION, &str));`